File: algorithms/04_java/math/CRT.java
public class CRT {
    // Solves x = rem[i] (mod mod[i]) for pairwise coprime moduli.
    public static long crt(long[] rem, long[] mod) {
        long prod = 1;
        for (long m : mod) prod *= m;
        long result = 0;
        for (int i = 0; i < rem.length; i++) {
            long pp = prod / mod[i];
            result += rem[i] * modInverse(pp, mod[i]) * pp;
            result %= prod;
        }
        return (result + prod) % prod;
    }

    private static long modInverse(long a, long m) {
        long[] r = extendedEuclid(a, m);
        if (r[0] != 1) throw new ArithmeticException("inverse does not exist");
        return (r[1] % m + m) % m;
    }

    private static long[] extendedEuclid(long a, long b) {
        if (b == 0) return new long[]{a, 1, 0};
        long[] r = extendedEuclid(b, a % b);
        return new long[]{r[0], r[2], r[1] - (a / b) * r[2]};
    }

    public static void main(String[] args) {
        long[] rem = {2, 3, 2};
        long[] mod = {3, 5, 7};
        long x = crt(rem, mod);
        if (x != 23) throw new AssertionError("CRT solution should be 23, got " + x);
        for (int i = 0; i < rem.length; i++)
            if (x % mod[i] != rem[i]) throw new AssertionError("CRT wrong for modulus " + mod[i]);
        System.out.println("[Java CRT] Chinese Remainder Theorem verified: x = " + x);
    }
}

File: algorithms/03_cpp/math/crt.cpp
/**
 * Chinese Remainder Theorem: solve x = r_i (mod m_i) for pairwise coprime moduli.
 */

#include <vector>
#include <cassert>
#include <iostream>

long long extgcd(long long a, long long b, long long& x, long long& y) {
    if (b == 0) {
        x = 1;
        y = 0;
        return a;
    }
    long long x1, y1;
    long long g = extgcd(b, a % b, x1, y1);
    x = y1;
    y = x1 - (a / b) * y1;
    return g;
}

long long modInverse(long long a, long long m) {
    long long x, y;
    extgcd(a, m, x, y);
    x %= m;
    if (x < 0) x += m;
    return x;
}

// Pairwise coprime moduli. Returns x mod (product of moduli).
long long crt(const std::vector<long long>& rem, const std::vector<long long>& mod) {
    long long x = 0;
    long long M = 1;
    for (size_t i = 0; i < mod.size(); ++i) {
        long long m = mod[i];
        long long r = ((rem[i] % m) + m) % m;
        // Solve M * t == r - x (mod m). gcd(M, m) == 1 since moduli are pairwise coprime.
        long long a = M % m;
        long long b = ((r - x) % m + m) % m;
        long long t = b * modInverse(a, m) % m;
        x += M * t;
        M *= m;
        x %= M;
        if (x < 0) x += M;
    }
    return x;
}

int main() {
    auto x = crt({2, 3, 2}, {3, 5, 7});
    assert(x == 23); // 23 % 3 == 2, 23 % 5 == 3, 23 % 7 == 2
    auto y = crt({1, 2, 3}, {2, 3, 5});
    assert(y == 23);
    std::cout << "[C++ CRT] Chinese Remainder Theorem verified." << std::endl;
    return 0;
}
